src/goto-instrument/loop_to_recursion.cpp

```diff
@@ -0,0 +1,858 @@
+  
+{
+   
+   goto_functionst::function_mapt::iterator rec_func=
+    goto_functions.function_map.find("REC_function");
+
+   //to specify loop locaion 
+   unsigned loop_begin_loc =0;
+   unsigned loop_end_loc =0;
+
+   goto_programt::targett loop_head;
+  
+   symbolt &rec_f = symbol_table.lookup("REC_function"); 
+  
+   /*before adding replacing loop by REC_function call, 
+   need to a)get arguments, b)set parameters, and c)set return*/
+   exprt::operandst arguments;
+   get_rec_function_arguments(goto_function, symbol_table, arguments);
+   
+   //should be done now for later use
+   std::map<exprt,symbol_exprt> recf_vars;
+ 
+   if(rec_func->second.type.parameters().empty())
+     {
+	set_rec_function_parameters(
+              rec_func->second,  
+              symbol_table,  
+              arguments, 
+              recf_vars);
+     }
+   
+   exprt lhs;
+   set_rec_function_return_expr(goto_functions, symbol_table, lhs);
+
+
+   Forall_goto_program_instructions(it, goto_function.body)
+   {
+
+    if(!it->is_backwards_goto() &&
+	it!=goto_function.body.instructions.end())
+    	continue; 
+
+
+    loop_head=it->get_target();
+    
+    //call code:
+    code_function_callt rec_function_call;
+    rec_function_call.add_source_location()=loop_head->source_location;
+    rec_function_call.lhs()=lhs;
+    rec_function_call.function()=rec_f.symbol_expr();
+    rec_function_call.arguments()= arguments;
+
+
+    // inser REC_function call:
+    goto_programt::targett rec_call_t=goto_function.body.insert_before(loop_head);
+
+    rec_call_t->make_function_call(rec_function_call);
+    rec_call_t->source_location=loop_head->source_location;
+    rec_call_t->code = rec_function_call;
+
+    //return value decl
+    goto_programt::targett rv_decl=goto_function.body.insert_before(rec_call_t);
+    rv_decl->make_decl();
+    rv_decl-> code =code_declt(lhs);
+    
+    //might need to re-assign struct components!
+ 
+    goto_function.body.update();
+
+    //loop location info 
+    loop_begin_loc = loop_head->location_number;
+    loop_end_loc = it->location_number;
+    
+   }
+
+   //ready to build REC_function body
+   build_rec_function_body( 
+        goto_functions,
+        goto_function,
+        symbol_table, 
+        loop_begin_loc, 
+        loop_end_loc,
+        arguments,
+        lhs,
+        recf_vars);
+}
+
+void loop_to_recursion(
+  goto_functionst &goto_functions, 
+  symbol_tablet &symbol_table,
+  message_handlert &message_handler)
+{
+
+  messaget message(message_handler);
+
+  bool loop_present = false; 
+
+  message.status() << "Searching for loops"
+                     << messaget::eom;
+ 
+  Forall_goto_functions(f_it, goto_functions)
+   {
+
+     const irep_idt function_identifier=f_it->first;
+     assert(!function_identifier.empty());
+     goto_functionst::goto_functiont &goto_function=f_it->second;
+
+     if(!goto_function.body_available())
+        continue;
+
+     if(has_loop(goto_function.body))
```

Comment:
In the end you should replace all the loops in all the functions by its own corresponding recursive function.